### PR TITLE
[CAPI] Propose dataset api sets

### PR DESCRIPTION
**Pull request by zhoonit**
- [CAPI] Propose dataset api sets

```
**Changes proposed in this PR:**
- `ml_train_dataset_create(ml_train_dataset_h *dataset);
- `ml_train_dataset_add_generator(dataset, usage, callback, user_data)`
- `ml_train_dataset_add_file(dataset, usage, file)`
- `ml_train_dataset_set_property_for_usage(dataset, usage)`

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: api/nntrainer-api-common.h [hunk 1]
```diff
@@ -151,6 +151,19 @@ typedef enum {
 typedef int (*ml_train_datagen_cb)(float **input, float **label, bool *last,
                                    void *user_data);
 
+/**
+ * @brief Enumeration for the dataset data type of NNTrainer.
```
Comment (kparichay) [inline api/nntrainer-api-common.h:155]: add more comments about the requirements of dataset - all of them are expected to provide input and label

File: api/capi/include/nntrainer.h [hunk 2]
```diff
@@ -358,6 +358,56 @@ int ml_train_dataset_create_with_generator(ml_train_dataset_h *dataset,
...
+ * saved to media storage.
+ * @remarks %http://tizen.org/privilege/externalstorage is needed if @a model is
+ * saved to external storage.
+ * @param[in] dataset The NNTrainer dataset handle.
+ * @param[in] usage The phase where this file should be used.
+ * @param[in] file file path.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #ML_ERROR_INVALID_PARAMETER Invalid parameter.
```
Comment (kparichay) [inline api/capi/include/nntrainer.h:405]: if the dataset already exist, the behavior is to overwrite? if yes, add as a comment here as well as in cb.

File: api/nntrainer-api-common.h [hunk 3]
```diff
@@ -151,6 +151,19 @@ typedef enum {
...
+ * @since_tizen 6.5
+ */
+typedef enum {
+  ML_TRAIN_DATASET_DATA_USAGE_TRAIN =
```
Comment (kparichay) [inline api/nntrainer-api-common.h:159]: ML_TRAIN_DATASET_TYPE_TRAIN?

**Review by kparichay — APPROVED** [review]
Overall LGTM other than some minor comments

File: api/nntrainer-api-common.h (hunk 3 above)
Comment (zhoonit) [reply api/nntrainer-api-common.h:159]: Would there be alternatives? I think dataset_type will be used later as 'file, image_directory' kind of stuff

File: api/capi/include/nntrainer.h (hunk 2 above)
Comment (zhoonit) [reply api/capi/include/nntrainer.h:405]: dataset_usage train exist, it will be error. I'll make a comment here.

File: api/nntrainer-api-common.h (hunk 1 above)
Comment (zhoonit) [reply api/nntrainer-api-common.h:155]: I think this is not a good place to add requirements, maybe dataset handle? Apart from it, I am planning to enhance tizen documentation regarding this topic.

File: api/capi/include/nntrainer.h (hunk 2 above)
Comment (kparichay) [reply api/capi/include/nntrainer.h:405]: so how does one change dataset set to a model. Does model allow setting a new dataset (replacing exist dataset)?

File: api/nntrainer-api-common.h (hunk 3 above)
Comment (kparichay) [reply api/nntrainer-api-common.h:159]: Isn't that data type? USAGE just sounds strange but works for now.

File: api/capi/include/nntrainer.h (hunk 2 above)
Comment (zhoonit) [reply api/capi/include/nntrainer.h:405]: setting dataset to a model is described in `ml_train_model_set_dataset()`. 

```
 * @remarks Unsets the previously set dataset, if any. The previously set
 * dataset must be freed using ml_train_dataset_destroy().
```

File: api/nntrainer-api-common.h (hunk 3 above)
Comment (zhoonit) [reply api/nntrainer-api-common.h:159]: Please throw me some word :p

`ml_train_dataset_add_generator(dataset, train /**<-usage */, cb, user_data)` will add generator to be used for training.

File: api/nntrainer-api-common.h (hunk 1 above)
Comment (kparichay) [reply api/nntrainer-api-common.h:155]: Which ever place seems good to you.

File: api/capi/include/nntrainer.h (hunk 2 above)
Comment (kparichay) [reply api/capi/include/nntrainer.h:405]: cool then.

File: api/nntrainer-api-common.h (hunk 3 above)
Comment (jijoongmoon) [reply api/nntrainer-api-common.h:159]: ML_TRAIN_DATASET_FOR_TRAINING.. it sounds strange.. ML_**TRAIN**_DATASET_DATA_FOR_**TRAIN**...
VALIDATING.. TESTING.. Maybe?

**Review by jijoongmoon — APPROVED** [review]
LGTM with mior comment

**Comment by jijoongmoon** [comment]
I wonder we discuss that previous api which need to be deprecated. I think we also need to add the notification in that api as well.

File: api/nntrainer-api-common.h (hunk 3 above)
Comment (zhoonit) [reply api/nntrainer-api-common.h:159]: I think it good idae!
As there are number of changes will break in the upcoming PR, I'll do this before making acr :)

Comment (zhoonit) [reply api/nntrainer-api-common.h:159]: As per offline discussion, this will be
```
typedef enum {
  ML_TRAIN_DATASET_MODE_TRAIN =
    0, /**< The given data is for used when training */
  ML_TRAIN_DATASET_MODE_VALID =
    1, /**< The given data is for used when validating */
  ML_TRAIN_DATASET_MODE_TEST =
    2, /**< The given data is for used when testing */
} ml_train_dataset_mode_e;
```

File: api/capi/include/nntrainer.h [hunk 4]
```diff
@@ -390,6 +390,56 @@ int ml_train_dataset_create_with_generator(ml_train_dataset_h *dataset,
...
+ */
+int ml_train_dataset_add_generator(ml_train_dataset_h dataset,
+                                   ml_train_dataset_data_usage_e usage,
+                                   ml_train_datagen_cb cb, void *user_data);
+
+/**
+ * @brief Add data file to @a dataset
+ * @details Use this function to add a data file from where data is retrieved.
+ * @remarks %http://tizen.org/privilege/mediastorage is needed if @a dataset is
```
Comment (p-czaja) [inline api/capi/include/nntrainer.h:427]: Please use `privlevel` tag instead.

File: api/capi/include/nntrainer.h [hunk 5]
```diff
@@ -390,6 +390,56 @@ int ml_train_dataset_create_with_generator(ml_train_dataset_h *dataset,
...
+ * saved to media storage.
+ * @remarks %http://tizen.org/privilege/externalstorage is needed if @a model is
```
Comment (p-czaja) [inline api/capi/include/nntrainer.h:429]: Please use `privlevel` tag instead.

File: api/capi/include/nntrainer.h (hunk 4 above)
Comment (zhoonit) [reply api/capi/include/nntrainer.h:427]: maybe previliege? privlevel seems to use for other purpose

Comment (p-czaja) [reply api/capi/include/nntrainer.h:427]: Sure. Actually: `@privilege` (not @previliege)
